src/platform/audio/alaudiomanager.cpp

```diff
@@ -0,0 +1,1587 @@
+
+    if (handle != nullptr) {
+        auto openal = audio->openal;
+        alSource3f(openal.source, AL_POSITION, pos->x, pos->y, pos->z);
+    }
+
+    return handle;
+}
+
+/**
+ * Loads a file an attached it to an OpenAL source. Returns a pointer to the file data.
+ */
+AudioDataHandle ALAudioManager::Play_Sample(AudioEventRTS *event, PlayingAudio *audio)
+{
+    AudioDataHandle handle = Open_File(event);
+
+    if (handle != nullptr) {
+        uint8_t *data = nullptr;
+        uint32_t size = 0, freq = 0;
+        uint8_t channels = 1, bits_per_sample = 16;
+#if BUILD_WITH_FFMPEG
+        Thyme::FFmpegAudioFileCache::Get_Wave_Data(handle, data, size, freq, channels, bits_per_sample);
+#endif
+        auto openal = audio->openal;
+        alGenBuffers(1, &openal.buffer);
+        alBufferData(openal.buffer, Get_AL_Format(channels, bits_per_sample), data, size, freq);
+        alSourcei(openal.source, AL_BUFFER, openal.buffer);
+        alSourcePlay(openal.source);
+    }
+
+    return handle;
+}
+
+/**
+ * Stops all speech playing audio.
+ */
+void ALAudioManager::Stop_All_Speech()
+{
+    auto it = m_streamList.begin();
+
+    while (it != m_streamList.end()) {
+        if (*it != nullptr) {
+            if ((*it)->openal.audio_event->Get_Event_Info()->Get_Event_Type() == EVENT_SPEECH) {
+                Release_Playing_Audio(*it);
+                it = m_streamList.erase(it);
+            } else {
+                ++it;
+            }
+        }
+    }
+}
+
+/**
+ * Attempts to start another loop of a playing audio file, return indicates success.
+ */
+bool ALAudioManager::Start_Next_Loop(PlayingAudio *audio)
+{
+    // Close_File(audio->openal.file_handle);
+    audio->openal.file_handle = nullptr;
+
+    if (audio->openal.disable_loops || !audio->openal.audio_event->Has_More_Loops()) {
+        return false;
+    }
+
+    audio->openal.audio_event->Generate_Filename();
+
+    if (audio->openal.audio_event->Get_Delay() > 33.333333f) {
+        audio->openal.release_event = false;
+        audio->openal.disable_loops = true;
+        audio->openal.stopped = 1;
+        AudioRequest *request = Allocate_Audio_Request(true);
+        request->m_event.object = audio->openal.audio_event;
+        request->m_isProcessed = true;
+        Append_Audio_Request(request);
+
+        return true;
+    }
+
+    if (audio->openal.playing_type == PAT_3DSAMPLE) {
+        audio->openal.file_handle = Play_Sample3D(audio->openal.audio_event, audio);
+    } else {
+        audio->openal.file_handle = Play_Sample(audio->openal.audio_event, audio);
+    }
+
+    return audio->openal.file_handle != nullptr;
+}
+
+/**
+ * Adjusts the volume of a playing audio object at the Miles audio level.
+ */
+void ALAudioManager::Adjust_Playing_Volume(PlayingAudio *audio)
+{
+    float adjusted_vol = audio->openal.audio_event->Get_Volume() * audio->openal.audio_event->Get_Volume_Shift();
+    alSourcef(audio->openal.source, AL_GAIN, adjusted_vol);
```

Comment:
This looks incomplete to me, because it does not incorporate `m_musicVolume` and `m_speechVolume` for example.